src/features/courses/course-route.ts

```diff
@@ -312,6 +313,36 @@ router.put('/topic/:id',
         }
     }));
 
+router.get('/assessment/topic/end/:id',
+    authenticationMiddleware,
+    validate(endAssessmentVersionValidation),
+    // This is due to a typescript issue where the type mismatches extractMap
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    asyncHandler(async (req: RederlyExpressRequest<any, unknown, EndAssessmentVersionRequest.body, EndAssessmentVersionRequest.query>, _res: Response, next: NextFunction) => {
+        const params = req.params as EndAssessmentVersionRequest.params;
+        if (_.isNil(req.session)) {
+            throw new Error(Constants.ErrorMessage.NIL_SESSION_MESSAGE);
+        }
+        const user = await req.session.getUser();
+        const version = await courseController.getStudentTopicAssessmentInfoById(params.id);
+
+        if (version.userId !== user.id) {
+            throw new ForbiddenError('You may not end an exam that does not belong to you.');
+        }
+
+        if (version.numAttempts === 0) {
+            throw new ForbiddenError('You cannot end an exam without making at least one attempt.');
+        }
+
+        // if (version.isClean === false) {
```

Comment:
I would think ending an exam would submit it if there are changes

Reply:
I have not yet implemented isClean - but yes, this would be ideal